onnxruntime/core/optimizer/transformer_memcpy.cc

```diff
@@ -255,14 +255,8 @@ void TransformerMemcpyImpl::ProcessDefs(onnxruntime::Node& node, const KernelReg
         provider_output_defs_.insert(arg);
     }
   } else if (node_provider_type != kCudaExecutionProvider && node_provider_type != kTensorrtExecutionProvider &&
-             node_provider_type != kRocmExecutionProvider && node_provider_type != kMIGraphXExecutionProvider) {
-    // TODO: copy between devices? i.e. multiple GPUs
-    if (node_provider_type != onnxruntime::kCpuExecutionProvider &&
-        node_provider_type != onnxruntime::kVitisAIExecutionProvider &&
-        !node_provider_type.empty()) {
-      ORT_THROW("Execution type '", node_provider_type, "' doesn't support memcpy ");
-    }
```

Comment:
I believe this is safe.

The provider is only ever a GPU EP

https://github.com/microsoft/onnxruntime/blob/d7367653ab28045ab95bf784642c6b0e48975df4/onnxruntime/core/optimizer/transformer_memcpy.cc#L65-L69

This 'else if' branch is setting up the non-provider inputs/outputs.

And we only ever assign the current provider to the memcpy node.

https://github.com/microsoft/onnxruntime/blob/d7367653ab28045ab95bf784642c6b0e48975df4/onnxruntime/core/optimizer/transformer_memcpy.cc#L335

So a non-provider would never be assigned the memcpy node, making this check irrelevant.

Reply:
One additional note. If we do add an invalid memcpy node model loading will fail anyway during kernel lookup, so removing this check won't change the end result.